autoprecompiles/src/range_constraint_optimizer.rs

```diff
@@ -0,0 +1,146 @@
+use powdr_constraint_solver::inliner::DegreeBound;
+use powdr_constraint_solver::range_constraint::RangeConstraint;
+use powdr_constraint_solver::solver::{new_solver, Solver};
+use powdr_number::FieldElement;
+
+pub type RangeConstraintMap<T, V> = BTreeMap<GroupedExpression<T, V>, RangeConstraint<T>>;
+
+pub trait RangeConstraintHandler<T: FieldElement> {
+    /// If the bus interaction *only* enforces range constraints, returns them
+    /// as a map of expressions to range constraints.
+    /// For example, if a bus interaction takes two arguments `a` and `b` and
+    /// enforces the range constraints `0 <= a < 2^b`, it is *not* a pure range
+    /// constraint if both values are unknown (because the allowed values of `a`
+    /// depend on `b`), but it is a pure range constraint if `b` is known.
+    /// Any stateful bus interaction is not a pure range constraint.
+    fn pure_range_constraints<V: Ord + Clone + Eq + Display + Hash>(
+        &self,
+        bus_interaction: &BusInteraction<GroupedExpression<T, V>>,
+    ) -> Option<RangeConstraintMap<T, V>>;
+
+    /// Given a set of range constraints, returns a list of bus interactions
+    /// that implements them. The implementation is free to implement multiple
+    /// range constraints using a single bus interaction.
+    /// As all input range constraints are unconditional, the multiplicity of
+    /// the returned bus interactions should be 1.
+    fn batch_make_range_constraints<V: Ord + Clone + Eq + Display + Hash>(
+        &self,
+        range_constraints: RangeConstraintMap<T, V>,
+    ) -> Vec<BusInteraction<GroupedExpression<T, V>>>;
+}
+
+/// Optimizes range constraints, minimizing the number of bus interactions. This step removes range
+/// constraints that are already implied by existing constraints, and it implements bit constraints
+/// via polynomial constraints, if the degree bound allows.
+pub fn optimize_range_constraints<T: FieldElement, V: Ord + Clone + Hash + Eq + Display>(
+    mut system: ConstraintSystem<T, V>,
+    bus_interaction_handler: impl BusInteractionHandler<T> + RangeConstraintHandler<T> + Clone,
+    degree_bound: DegreeBound,
+) -> ConstraintSystem<T, V> {
+    // Remove all pure range constraints, but collect what was removed.
+    let mut to_constrain = BTreeMap::new();
+    system.bus_interactions.retain(|bus_int| {
+        if bus_int.multiplicity != GroupedExpression::from_number(T::one()) {
+            // Most range constraints are unconditional in practice, it's probably not
+            // worth dealing with the conditional ones.
+            return true;
+        }
+
+        match bus_interaction_handler.pure_range_constraints(bus_int) {
+            Some(range_constraints) => {
+                for (expr, rc) in range_constraints {
+                    let existing_rc = to_constrain
+                        .entry(expr)
+                        .or_insert_with(RangeConstraint::default);
```

Comment:
Not too important at this point, but having the expression as key of a map sounds rather expensive.